pkg/sync/rpc/server.go

```diff
@@ -202,26 +202,53 @@ func (s *SyncAgentServer) PrepareRestore(backupURL, requestedBackupName, snapsho
 		return fmt.Errorf("already restored backup %v", requestedBackupName)
 	}
 
+	// Initialize `s.RestoreInfo`
 	// First restore request. It must be a normal full restore.
 	if restoreStatus.LastRestored == "" && restoreStatus.State == "" {
 		s.RestoreInfo = replica.NewRestore(snapshotDiskName, s.replicaAddress, backupURL, requestedBackupName)
-		return nil
-	}
+	} else {
+		// By default this is a normal incremental restore.
+		toFileName := replica.GenerateDeltaFileName(restoreStatus.LastRestored)
+		validLastRestoredBackup := true
+
+		// Fall back to full restore when the field is invalid or the backup inspection fails.
+		if restoreStatus.LastRestored == "" {
+			logrus.Warnf("There is a restore record in the server but last restored backup is empty with restore state is %v, will do full restore instead", restoreStatus.State)
+			toFileName = replica.GenerateSnapTempFileName(snapshotDiskName)
+			validLastRestoredBackup = false
+		} else if _, err := backupstore.InspectBackup(strings.Replace(backupURL, requestedBackupName, restoreStatus.LastRestored, 1)); err != nil {
+			logrus.Warnf("The last restored backup %v becomes invalid for incremental restore, will do full restore instead, err: %v", restoreStatus.LastRestored, err)
+			toFileName = replica.GenerateSnapTempFileName(snapshotDiskName)
+			validLastRestoredBackup = false
+		}
 
-	// Fall back to full restore.
-	if restoreStatus.LastRestored == "" {
-		logrus.Warnf("There is a restore record in the server but last restored backup is empty with restore state is %v, will do full restore instead", restoreStatus.State)
-		s.RestoreInfo.StartNewRestore(backupURL, requestedBackupName, replica.GenerateSnapTempFileName(snapshotDiskName), snapshotDiskName, false)
-		return nil
+		s.RestoreInfo.StartNewRestore(backupURL, requestedBackupName, toFileName, snapshotDiskName, validLastRestoredBackup)
 	}
```

Comment:
How about adding a private util function for the condition checking? 
This makes the code easier to follow without losing any of the details.

example below:
```
isIncrementalRestore := canDoIncrementalRestore(x,y,z)
var toFilename
if isIncrementalRestore {
    toFileName = replica.GenerateDeltaFileName(restoreStatus.LastRestored)
 } else {
     toFileName = replica.GenerateSnapTempFileName(snapshotDiskName)
 }

s.RestoreInfo.StartNewRestore(backupURL, requestedBackupName, toFileName, snapshotDiskName, isIncrementalRestore)
```